extra error message if no command line arguments

mschwamb opened the pull request:
Fixes #1019

Extra error message if no command line arguments.

## Review Checklist for Source Code Changes

- [X] Does pip install still work?
- [X] Have you written a unit test for any new functions?
- [X] Do all the units tests run successfully?
- [X] Does Sorcha run successfully on a test set of input files/databases?
- [X] Have you used black on the files you have updated to confirm python programming style guide enforcement?

mschwamb commented:
Since I'm mucking with the CLI @mjuric  can you take a look? Should take you 2 minutes. Some people using the CLI thought it wasn't clear what to do if you ran sorcha-run with no arguments. So I added an extra message to STDERR if there's no user provided arguments.

mjuric commented on src/sorcha_cmdline/run.py:
```diff
@@ -116,6 +118,16 @@ def main():
+    if len(sys.argv) == 1:
+        print(
+            "\n\nError: No command line arguments provided!!! \nUse the -h flag to see descriptions of the required and optional arguments needed to run Sorcha \n",
+            file=sys.stderr,
```
I'd propose to do something like this here:

```
parser.print_usage(file=sys.stderr)
print("For more detailed help run: sorcha run --help", file=sys.stderr)
```

A better soluton (which would solve this for all subcommends) is to derive a class from`ArgumentParser` and override the `print_usage` function. Example:

```
class SorchaArgumentParser(argparse.ArgumentParser):
    def print_usage(self, file=None):
        super().print_usage(file)
        print("For more detailed help run: sorcha run --help", file=file)
```

and then just change line 9 to `parser = SorchaArgumentParser`.

(you'd probably want to put SorchaArgumentParser in some util module in `sorcha_cmdline` package).

mjuric replied:
Correction, you'll need something like this:
```
class SorchaArgumentParser(argparse.ArgumentParser):
    def __init__(self, cmdname, *args, **kwargs):
        self.cmdname = cmdname
        return super().__init__(**args, **kwargs)

    def print_usage(self, file=None):
        super().print_usage(file)
        print(f"For more detailed help run: sorcha {self.cmdname} --help", file=file)
```
so as not to hardcode the subcommand name.

mschwamb commented:
I even know what you're talking about. So maybe I did learn something python class stuff. I agree, your way is more elegant. I was thinking we didn't need to do this for all the times that someone screwed up the arguments, but I see your way is classier. I'll see where I get to and request a review when I've implemented this. Though it's gonna take me a bit to look at the python class tutorials again for me to do this/ double check what you've written will work (it probably will)

mjuric commented:
FWIW, your current solution is also OK, if you want to quickly merge this PR & move the generalization to "future enhancements".

I'd offer to implement this myself but I'm underwater with return travel & other things for a few more days :(.

mschwamb commented:
I did some reading implemented this and I'm getting this behavior that works but the output is a bit odd:

usage: ArgumentParser(prog='sorcha-run', usage=None, description='Run a simulation.', formatter_class=<class 'argparse.ArgumentDefaultsHelpFormatter'>, conflict_handler='error', add_help=True)
       [-h] -c C -o O -ob OB -p P -pd PD [-er ER] [-ew EW] [-ar AR] [-cp CP] [-f] [-s S] [-t T] [-v] [-st ST]
For more detailed help, use the -h flag
ArgumentParser(prog='sorcha-run', usage=None, description='Run a simulation.', formatter_class=<class 'argparse.ArgumentDefaultsHelpFormatter'>, conflict_handler='error', add_help=True): error: the following arguments are required: -c/--config, -o/--outfile, -ob/--orbit, -p/--params, -pd/--pointing_database
Error: Command 'sorcha-run' failed with exit code 2.

instead of 

usage: sorcha-run [-h] -c C -o O -ob OB -p P -pd PD [-er ER] [-ew EW] [-ar AR] [-cp CP] [-f] [-s S] [-t T] [-v] [-st ST]
For more detailed help, use the -h flag
sorcha-run: error: the following arguments are required: -c/--config, -o/--outfile, -ob/--orbit, -p/--params, -pd/--pointing_database
Error: Command 'sorcha-run' failed with exit code 2.

I implemented 
```

class sargparse(argparse.ArgumentParser):
    def __init__(self, *argv, **kwargs):
       super().__init__(*argv, **kwargs)
    
 
    def print_usage(self,file=None):
        super().print_usage(file)
        print("For more detailed help, use the -h flag",file=sys.stderr)
   
    def main():
    parser = argparse.ArgumentParser(
        formatter_class=argparse.ArgumentDefaultsHelpFormatter, description="Run a simulation."
    )
    
    parser=sargparse(parser)
    .......
```
        
I think this change makes it more confusing with the "usage: ArgumentParser(prog='sorcha-run', usage=None, description='Run a simulation.', formatter_class=<class 'argparse.ArgumentDefaultsHelpFormatter'>, conflict_handler='error', add_help=True)"  bit. If you have ideas @mjuric how to get around this let me know. Otherwise I think it's easier to do what I implemented though I'll short the message to what you've written.

mschwamb commented:
I figured it out. I'll get the PR in, in about half an hour.
